my-sms/spec/controllers/students_controller_spec.rb

```diff
@@ -23,119 +25,115 @@
-        }.to change(Student, :count).by(1)
+  describe 'POST #create' do
+    subject { post :create, { student: student } }
+
+    context 'with valid params' do
+      let(:student) { FactoryBot.attributes_for(:student) }
+
+      it 'creates a new Student' do
+        expect { subject }.to change(Student, :count).by(1)
       end
 
-      it "redirects to the created student" do
-        post :create, {:student => valid_attributes}, valid_session
-        expect(response).to redirect_to(Student.last)
+      it 'redirects to the created student' do
+        should be_redirect
+        should redirect_to(Student.last)
       end
     end
 
-    context "with invalid params" do
+    context 'with invalid params' do
+      let(:student) { FactoryBot.attributes_for(:student, first_name: nil) }
+
       it "returns a success response (i.e. to display the 'new' template)" do
-        post :create, {:student => invalid_attributes}, valid_session
-        expect(response).to be_success
+        should be_success
+        should render_template('new')
       end
     end
   end
 
-  describe "PUT #update" do
-    context "with valid params" do
-      let(:new_attributes) {
-        skip("Add a hash of attributes valid for your model")
-      }
+  describe 'PUT #update' do
+    let(:student) { FactoryBot.create(:student) }
+    subject { put :update, { id: student.to_param, student: new_attributes } }
+
+    context 'with valid params' do
+      let(:new_attributes) { FactoryBot.attributes_for(:student) }
 
-      it "updates the requested student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => new_attributes}, valid_session
+      it 'updates the requested student' do
+        subject
         student.reload
-        skip("Add assertions for updated state")
+        attrs = %i[first_name last_name gender]
+        student_attrs = student.attributes.fetch_values(*attrs.map(&:to_s))
+        new_attrs = new_attributes.fetch_values(*attrs)
+        expect(student_attrs).to eq new_attrs
       end
 
-      it "redirects to the student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => valid_attributes}, valid_session
-        expect(response).to redirect_to(student)
+      it 'redirects to the student' do
+        should be_redirect
+        should redirect_to(student)
       end
     end
 
-    context "with invalid params" do
+    context 'with invalid params' do
+      let(:new_student_args) { { first_name: nil } }
+      let(:new_attributes) do
+        FactoryBot.attributes_for(:student, { first_name: nil })
+      end
+
       it "returns a success response (i.e. to display the 'edit' template)" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => invalid_attributes}, valid_session
-        expect(response).to be_success
+        should be_success
```

Comment:
Also consider testing that no new record is created:

```ruby
it 'does not create a new record'
  expect { subject }.to_not change { Student.count }
end
```

Reply:
Thanks, added in d1c657a7ddbcc2e460c38ad78d1400a7799c5297